cmd/geth/import_from_firehose.go

```diff
@@ -0,0 +1,306 @@
+	}
+
+	apiToken := os.Getenv("FIREHOSE_API_TOKEN")
+	endpoint := ctx.Args().Get(0)
+	chainIDStr := ctx.Args().Get(1)
+	externalRpc := ctx.Args().Get(2)
+	batchSize := ctx.Int("batch-size")
+	endBlock := ctx.Uint64("end-block")
+	workerCount := ctx.Int("worker-count")
+	bufferSize := ctx.Int("firehose-buffer-size")
+
+	chainID := new(big.Int)
+	if _, ok := chainID.SetString(chainIDStr, 10); !ok {
+		return fmt.Errorf("invalid chainID: %s", chainIDStr)
+	}
+
+	// Open Geth stack and chain
+	stack, cfg := makeConfigNode(ctx)
+	defer stack.Close()
+	utils.SetupMetrics(&cfg.Metrics)
+	chain, db := utils.MakeChain(ctx, stack, false)
+	defer db.Close()
+	defer chain.Stop()
+
+	var startBlock int
+	if ctx.IsSet("start-block") {
+		startBlock = ctx.Int("start-block")
+		if startBlock < 0 {
+			return fmt.Errorf("startBlock must be non-negative")
+		}
+		fmt.Printf("Starting from user-specified block: %d\n", startBlock)
+	} else {
+		// Resume from the local chain head + 1
+		head := chain.CurrentBlock()
+		if head != nil {
+			startBlock = int(head.Number.Uint64() + 1)
+			fmt.Printf("No start block specified. Resuming from last imported block: %d\n", startBlock)
+		} else {
+			startBlock = 0 // start from genesis
+			fmt.Println("No start block specified and no local chain found. Starting from genesis (block 0).")
+		}
+	}
+
+	var totalBlocks int
+	currentBlock := startBlock
+	err := processFirehoseBlocksWithReconnect(endpoint, apiToken, &currentBlock, endBlock, batchSize, workerCount, bufferSize, chainID, externalRpc, func(blocks []*types.Block, batchNum int) error {
+		if len(blocks) == 0 {
+			return nil
+		}
+		firstNum := blocks[0].NumberU64()
+		lastNum := blocks[len(blocks)-1].NumberU64()
+		if _, err := chain.InsertChain(blocks); err != nil {
+			fmt.Printf("failed to import batch %d (blocks %d-%d): %v\n", batchNum, firstNum, lastNum, err)
+			return err
+		}
+		fmt.Printf("Imported batch %d of %d blocks (blocks %d-%d)\n", batchNum, len(blocks), firstNum, lastNum)
+		totalBlocks += len(blocks)
+		currentBlock = int(lastNum + 1)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Import completed successfully. Total blocks imported: %d\n", totalBlocks)
+	return nil
+}
+
+func processFirehoseBlocksWithReconnect(
+	endpoint string,
+	apiToken string,
+	startBlock *int,
+	endBlock uint64,
+	batchSize int,
+	workerCount int,
+	bufferSize int,
+	chainID *big.Int,
+	externalRpc string,
+	handler func(blocks []*types.Block, batchNum int) error,
+) error {
+	maxRetries := 100
+	attempt := 0
+
+	for attempt < maxRetries {
+		prev := *startBlock
+
+		err := retry.Do(
+			func() error {
+				return processFirehoseBlocks(endpoint, apiToken, startBlock, endBlock, batchSize, workerCount, bufferSize, chainID, externalRpc, handler)
+			},
+			retry.Attempts(1),
+			retry.DelayType(retry.BackOffDelay),
+			retry.Delay(time.Second*2),
+			retry.MaxDelay(time.Minute*5),
+		)
```

Comment:
Make it forever repeating, unless the error is gRPC Unauthenticated or InvalidArgument, those are not retryable. You can check here how we deal with retryable errors:

- https://github.com/streamingfast/substreams-sink/blob/develop/sinker.go#L390-L401
- https://github.com/streamingfast/substreams-sink/blob/develop/sinker.go#L323-L337